src/text/scrptrun.cpp

```diff
@@ -156,7 +161,8 @@ UBool ScriptRun::next()
         // characters above it on the stack will be poped.
         if (pairIndex >= 0) {
             if ((pairIndex & 1) == 0) {
-                parenStack[++parenSP].pairIndex = pairIndex;
+                parenSP = (++parenSP) & STACK_MASK; // avoid out-of-bounds access
```

Comment:
This has a preincrement and an assignment in the same expression. I think this instead needs to be

```
parenSP = (parenSP + 1) & STACK_MASK;
```

With this change, it looks like this should prevent the crash, since the corresponding code that decrements `parenSP` checks for 0 instead of blindly decrementing.

It will still have strange behavior when punctuation is nested this deeply, since the punctuation will no longer be treated as paired, but incorrect rendering is a lot better than crashing. I don't know if it would be practical from a performance perspective to use a dynamically-allocated vector for the stack instead of an array, but if it is fast enough, that would ultimately be better.

Reply:
> This has a preincrement and an assignment in the same expression. I think this instead needs to be
> 
> ```
> parenSP = (parenSP + 1) & STACK_MASK;
> ```
yep, fixed, cheers!

> With this change, it looks like this should prevent the crash, since the corresponding code that decrements `parenSP` checks for 0 instead of blindly decrementing.
> 
> It will still have strange behavior when punctuation is nested this deeply, since the punctuation will no longer be treated as paired, but incorrect rendering is a lot better than crashing. I don't know if it would be practical from a performance perspective to use a dynamically-allocated vector for the stack instead of an array, but if it is fast enough, that would ultimately be better.

This is rather unusual situation to have > 128 nested paired punctuation chars:). Also I thought about not diverging too far from the original ICU code (for future references). But yes, I'm thinking about std::vector too.